packages/apollo-server-core/src/types.ts

```diff
@@ -124,6 +124,7 @@ export interface Config extends BaseConfig {
   playground?: PlaygroundConfig;
   gateway?: GraphQLService;
   experimental_approximateDocumentStoreMiB?: number;
+  handleSignals?: boolean;
```

Comment:
Keeping in mind all the other signals that we do not handle, I don't know that this name is clear enough at the top-level.  It's really more like "handle TERM and INT on Node.js by calling AS's stop method".

Is it critical to ship the signal trapping as part of this implementation?

There was a very real reason why I have mentioned trying to avoid taking an opinionated stance as to what constitutes a shutdown for Apollo Server and self-instrumenting that since `process.on` is a very Node.js-specific runtime concern.  Further, `SIGTERM` and `SIGINT` both have default handlers in Node.js that I believe disappear when we provide our own.

Thus far, I've been fine with `apollo-engine-reporting` using `process`-things and performing this shutdown since it also uses other Node.js specific bits (like `process.hrtime`) which render it unusable on non-Node.js runtimes.  But hoisting this makes me want to think about it a bit more.  For example, I'm not sure what happens on Windows, and Deno — if we ever can ever support it — uses a different mechanism (though I believe we could feature check for its `process`-equivalent.

I'm not sure we really know what other existing configuration or races against `process.on` handlers we're putting ourselves in the queue against.  I believe those other mechanisms can short-circuit or trap the server's actual shutdown which might leave us running without the polling?

In my mind, to avoid these concerns, it seems plausible that it's better to leave it in the hands of the user and guide/document them accordingly.  In Apollo Server 3.x's single package model, we still need to consider what the server lifetime hooks are.  For example, maybe there's a different approach where we don't have a server lifetime, but rather use `setTimeout` (not `setInterval`!) to schedule work that is intentionally keeping the event loop alive, but not create agent-like-things that need to be stopped?

Reply:
I do think that stripping a final semi-reliable flush from reporting will have a negative impact on the quality of using Studio, and that requiring it to be 100% opt-in will have a negative impact too. This is going to affect any plugin that's batching information to be sent to any sort of monitoring service, so I think any AS3 plan has to have some solution for this use case.

We've had the AER signal handling on by default for over two years; I'm curious if you've heard any user complaints? I do agree that it doesn't feel bulletproof, but if we haven't had any complaints yet, then it feels "good enough for AS2" to me.

Certainly having it only default to true in Node-like contexts makes sense.

What if the default is "true if Node and Unix (how do I detect this?) *and some plugin implements serverWillStop*"?

Reply:
Oh I certainly can agree re naming. `stopOnSIGINTAndSIGTERM`?

Reply:
How about `handleTerminationSignals`, `handleTermSignals`, `handleTerminationSignals`, or `stopOnTermSignals`?  Not sure if this is too ambiguous or overly verbose.  I was just thinking of trying to get the conjunction (e.g. "And") out of the variable name.

I drew inspiration from: https://www.gnu.org/software/libc/manual/html_node/Termination-Signals.html

Reply:
I like stopOnTerminationSignals.

Reply:
I don't feel great about my suggestion of "register signal handler if plugins are looking for it", because the gateway is not a plugin and I do want it to be reliably stopped when I implement #4428. I'm going to stick with "any Node-like non-test environment defaults to handling signals".

I'll note that I see zero references to `handleSignals` in issues on this repo or in our internal Slack which makes me believe that the current default behavior has not been a problem enough for users that we have ended up recommending "oh don't worry just set `handleSignals: false`" to anyone ever. Similarly looking at GitHub code search ([TypeScript](https://github.com/search?l=TypeScript&q=%22handleSignals%3A+false%22&type=Code), [JS](https://github.com/search?l=JavaScript&q=%22handleSignals%3A+false%22&type=Code)) doesn't show any public repos where people are setting this.

While the proposal certainly does broaden the contexts where the signals are handled, it seems like it hasn't been a problem so far. It also feels like a pretty reversible decision: if it causes a problem, then (a) there's the trivial workaround of setting `stopOnTerminationSignals: false` and (b) we could perhaps adjust the heuristic to be less aggressive about defaulting to true in a future release. However, my suspicion based on our two years of AER signal handling is that it will not be a problem.